docs/math/crt.md

```diff
@@ -81,9 +81,72 @@ $$
 4. 七子团圆正 **半月** ： $n_3=7, m_3=n/n_3=15, m_3^{-1}\equiv 1\pmod 7$ ，故 $c_3=15\times 1=15$ ；
 5. 所以方程组的唯一解为 $a\equiv 2\times 70+3\times 21+2\times 15\equiv 233\equiv 23 \pmod {105}$ 。（除 **百零五** 便得知）
 
+## Garner 算法
+
+CRT 的另一个用途是用一组比较小的整数表示一个大的整数。
+例如，令 $p$ 是前一千个素数的乘积，通过计算我们得知 $p$ 在十进制下大概有 3000 位。
+
+任意一个小于 $p$ 的数 $a$ 都能用一组数 $a_1, \ldots, a_k$ 表示，其中 $a_i \equiv a \pmod{p_i}$ 。但是要实现上述表示我们首先要知道如何从它的表示中算出 $a$ ，而 **Garner 算法** 就可以解决这个问题。我们用以下形式的式子（称作 $a$ 的混合基数表示）表示 $a$ ：
+
+$$
+    a = x_1 + x_2 p_1 + x_3 p_1 p_2 + \ldots + x_k p_1 \ldots p_{k-1}
+$$
+
+ **Garner 算法** 将用来计算系数 $x_1, \ldots, x_k$ .
+
+令 $r_{ij}$ 为 $p_i$ 模 $p_j$ 的逆：
+
+$$
+    r_{ij} = (p_i)^{-1} \pmod{p_j}
+$$
+
+求逆的算法可以在 [乘法逆元](./inverse.md) 中找到。把 $a$ 代入我们得到的第一个方程：
+
+$$
+    a_1 \equiv x_1 \pmod{p_1}.
+$$
+
+代入第二个方程得出：
+
+$$
+    a_2 \equiv x_1 + x_2 p_1 \pmod{p_2}.
+$$
+
+方程两边减 $x_1$ ，除 $p_1$ 后得
+
+$$
+    a_2 - x_1 &\equiv& x_2 p_1 &\pmod{p_2} \\\\
+    (a_2 - x_1) r_{12} &\equiv& x_2 &\pmod{p_2} \\\\
+    x_2 &\equiv& (a_2 - x_1) r_{12} &\pmod{p_2}
+\end{array}
+$$
+
+同样我们可以得到：
+
+$$
+    x_3 \equiv ((a_3 - x_1) r_{13} - x_2) r_{23} \pmod{p_3}.
+$$
+
+??? note "参考代码"
+    ```cpp
+    for (int i = 0; i < k; ++i) {
+      x[i] = a[i];
+      for (int j = 0; j < i; ++j) {
+        x[i] = r[j][i] * (x[i] - x[j]);
+    ```
+
+            x[i] = x[i] % p[i];
+            if (x[i] < 0)
+                x[i] += p[i];
+        }
+    }
+    ```
```

Comment:
```suggestion
??? note "参考代码"
    ```cpp
    for (int i = 0; i < k; ++i) {
      x[i] = a[i];
      for (int j = 0; j < i; ++j) {
        x[i] = r[j][i] * (x[i] - x[j]);
        x[i] = x[i] % p[i];
        if (x[i] < 0){
            x[i] += p[i];
        }
    }
    ```
```